zebra/zserv.c

```diff
@@ -435,21 +441,48 @@ static void zserv_read(struct event *thread)
 			while (cache->head)
 				stream_fifo_push(client->ibuf_fifo,
 						 stream_fifo_pop(cache));
+			/*Need to update count since main thread could have processed few*/
+			client_ibuf_fifo_cnt =
+				stream_fifo_count_safe(client->ibuf_fifo);
 		}
 
 		/* Schedule job to process those packets */
 		zserv_event(client, ZSERV_PROCESS_MESSAGES);
 
 	}
 
+    /*
+     * We arrive at this point in three cases.
+     * Case-1: Initially Client buffer is already to its max limit i.e.
+     *         client_ibuf_fifo == p2p_orig.
+     * Case-2: Initially Client buffer is Not at the max limit but becomes full
+     *         i.e. Initial client buffer had 90 elements with max (say 100).
+     *         So we add 10 more items to fifo and reach the max limit.
+     * Case-3: Initially Client buffer is Not at the max limit and neither is
+     *         it after the handling of new incoming buffer msg
+     *         i.e. Initial client buffer had 40 elements with max (say 100).
+     *         So now we add 20 more items to fifo.
+     *
+     * Only in case-3 we reschedule ourselves.
+     * In case-1 and case-2, the zserv_process_messages() (main thread)
+     * after processing the msgs from client input buffer (ibuf_fifo), wakes up
+     * the current client pthread to continue reading incoming messages.
+     */
 	if (IS_ZEBRA_DEBUG_PACKET)
-		zlog_debug("Read %d packets from client: %s", p2p_orig - p2p,
-			   zebra_route_string(client->proto));
-
-	/* Reschedule ourselves */
-	zserv_client_event(client, ZSERV_CLIENT_READ);
-
-	stream_fifo_free(cache);
+		zlog_debug("Read %d packets from client: %s. Current ibuf fifo count: %zu",
+			   p2p_orig - p2p, zebra_route_string(client->proto),
+			   client_ibuf_fifo_cnt);
+
+	if (client_ibuf_fifo_cnt == p2p_orig) {
+		/* Case-1, Do nothing */
+		/* Case-2*/
+		if (p2p != p2p_orig)
```

Comment:
hmm, we did a `stream_fifo_new()` above - don't we have to free that always?

Reply:
You are right.. wil fix this